File: apps/web/src/types/subscription.ts
// Subscription tier type
export type SubscriptionTier = "free" | "pro";
export type SubscriptionStatus = "active" | "expired" | "cancelled" | "none";
export type CodeType = "single_use" | "multi_use";
export type SubscriptionAction =
  | "created"
  | "upgraded"
  | "renewed"
  | "expired"
  | "cancelled"
  | "migrated";

// Subscription plan features
export interface SubscriptionFeatures {
  ai_conversations_per_day?: number; // -1 = unlimited
  context_window_tokens?: number;
  file_versions_per_file?: number;
  max_projects?: number; // -1 = unlimited
  export_formats?: string[];
  custom_prompts?: boolean;
  materials_library_access?: boolean;
  material_uploads?: number; // -1 = unlimited
  material_decompositions?: number;
  custom_skills?: number; // -1 = unlimited
  inspiration_copies_monthly?: number; // -1 = unlimited
  priority_support?: boolean;
  [key: string]: unknown;
}

// Subscription plan
export interface SubscriptionPlan {
  id: string;
  name: SubscriptionTier;
  display_name: string;
  display_name_en?: string;
  price_monthly_cents: number;
  price_yearly_cents: number;
  features: SubscriptionFeatures;
  is_active: boolean;
  created_at?: string;
  updated_at?: string;
}

export interface SubscriptionCatalogEntitlements {
  writing_credits_monthly: number;
  agent_runs_monthly: number;
  active_projects_limit: number;
  context_tokens_limit: number;
  materials_library_access: boolean;
  material_uploads_monthly: number;
  material_decompositions_monthly: number;
  custom_skills_limit: number;
  inspiration_copies_monthly: number;
  export_formats: string[];
  priority_queue_level: "standard" | "priority";
}

export interface SubscriptionCatalogTier {
  id: string;
  name: string;
  display_name: string;
  display_name_en?: string;
  price_monthly_cents: number;
  price_yearly_cents: number;
  recommended: boolean;
  summary_key: string;
  target_user_key: string;
  entitlements: SubscriptionCatalogEntitlements;
}

export interface SubscriptionCatalogResponse {
  version: string;
  comparison_mode: string;
  pricing_anchor_monthly_cents: number;
  tiers: SubscriptionCatalogTier[];
}

// User subscription
export interface UserSubscription {
  id: string;
  user_id: string;
  plan_id: string;
  plan?: SubscriptionPlan;
  status: SubscriptionStatus;
  current_period_start: string;
  current_period_end: string;
  cancel_at_period_end: boolean;
  created_at: string;
  updated_at: string;
}

// Usage quota
export interface UsageQuota {
  id: string;
  user_id: string;
  period_start: string;
  period_end: string;
  ai_conversations_used: number;
  last_reset_at: string;
  created_at: string;
  updated_at: string;
}

// Redemption code
export interface RedemptionCode {
  id: string;
  code: string;
  code_type: CodeType;
  tier: SubscriptionTier;
  duration_days: number;
  max_uses: number | null;
  current_uses: number;
  created_by: string;
  is_active: boolean;
  expires_at: string | null;
  notes: string | null;
  created_at: string;
  redeemed_by: string[];
}

// Subscription history
export interface SubscriptionHistory {
  id: string;
  user_id: string;
  action: SubscriptionAction;
  plan_name: string;
  start_date: string;
  end_date: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
}

// Admin audit log
export interface AdminAuditLog {
  id: string;
  admin_user_id: string;
  action: string;
  resource_type: string;
  resource_id: string | null;
  old_value: Record<string, unknown> | null;
  new_value: Record<string, unknown> | null;
  ip_address: string | null;
  user_agent: string | null;
  created_at: string;
}

// API Response types
export interface SubscriptionStatusResponse {
  tier: SubscriptionTier;
  status: SubscriptionStatus;
  display_name: string;
  display_name_en?: string | null;
  current_period_end: string | null;
  days_remaining: number | null;
  features: SubscriptionFeatures;
}

export interface QuotaResponse {
  ai_conversations: QuotaMetric;
  projects: QuotaMetric;
  material_uploads: QuotaMetric;
  material_decompositions: QuotaMetric;
  skill_creates: QuotaMetric;
  inspiration_copies: QuotaMetric;
}

export interface QuotaMetric {
  used: number;
  limit: number;
  reset_at: string | null;
}

export interface RedeemCodeRequest {
  code: string;
  source?: string;
}

export interface RedeemCodeResponse {
  success: boolean;
  message: string;
  tier?: SubscriptionTier;
  duration_days?: number;
  subscription?: UserSubscription;
}

File: apps/server/models/subscription.py
"""
Subscription models for user billing and quota management.

Defines SQLModel entities for:
- SubscriptionPlan: Available subscription tiers (e.g. free, pro, max)
- UserSubscription: User's active subscription
- RedemptionCode: Gift/redemption codes for subscription upgrades
- UsageQuota: Period-based usage tracking
- SubscriptionHistory: Subscription change history
- AdminAuditLog: Admin action audit trail
"""

from datetime import datetime

from sqlalchemy import JSON, Column
from sqlmodel import Field, SQLModel

from .utils import generate_uuid


class SubscriptionPlan(SQLModel, table=True):
    """
    Subscription plan model.

    Defines available subscription tiers with pricing and features.
    Tiers: "free", "pro" (extensible)
    """

    __tablename__ = "subscription_plan"

    id: str = Field(default_factory=generate_uuid, primary_key=True)
    name: str = Field(unique=True, index=True)  # e.g. "free", "pro"
    display_name: str
    display_name_en: str | None = None
    price_monthly_cents: int = 0
    price_yearly_cents: int = 0
    # Stripe fields (future)
    stripe_price_id_monthly: str | None = None
    stripe_price_id_yearly: str | None = None
    features: dict = Field(default_factory=dict, sa_column=Column(JSON))
    is_active: bool = True
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class UserSubscription(SQLModel, table=True):
    """
    User subscription model.

    Tracks each user's active subscription and billing period.
    One subscription per user (user_id is unique).
    """

    __tablename__ = "user_subscription"

    id: str = Field(default_factory=generate_uuid, primary_key=True)
    user_id: str = Field(foreign_key="user.id", unique=True, index=True)
    plan_id: str = Field(foreign_key="subscription_plan.id")
    status: str = "active"  # "active", "expired", "cancelled"
    current_period_start: datetime
    current_period_end: datetime
    cancel_at_period_end: bool = False
    # Stripe fields (future)
    stripe_subscription_id: str | None = None
    stripe_customer_id: str | None = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class RedemptionCode(SQLModel, table=True):
    """
    Redemption code model for subscription upgrades.

    Supports:
    - Single-use codes (one redemption)
    - Multi-use codes (max_uses limit)
    Codes can grant Pro tier access for a specified duration.
    """

    __tablename__ = "redemption_code"

    id: str = Field(default_factory=generate_uuid, primary_key=True)
    code: str = Field(unique=True, index=True)
    code_type: str = "single_use"  # "single_use", "multi_use"
    tier: str = Field(foreign_key="subscription_plan.name")
    duration_days: int
    max_uses: int | None = None
    current_uses: int = 0
    created_by: str = Field(foreign_key="user.id")
    is_active: bool = True
    expires_at: datetime | None = None
    notes: str | None = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    redeemed_by: list = Field(default_factory=list, sa_column=Column(JSON))


class UsageQuota(SQLModel, table=True):
    """
    Usage quota model for period-based tracking.

    Tracks AI conversation usage within a billing period.
    One quota per user (user_id is unique).
    """

    __tablename__ = "usage_quota"

    id: str = Field(default_factory=generate_uuid, primary_key=True)
    user_id: str = Field(foreign_key="user.id", unique=True, index=True)
    period_start: datetime
    period_end: datetime
    ai_conversations_used: int = 0
    last_reset_at: datetime = Field(default_factory=datetime.utcnow)
    # 月度功能使用量
    material_uploads_used: int = Field(default=0)
    material_decompositions_used: int = Field(default=0)
    skill_creates_used: int = Field(default=0)
    inspiration_copies_used: int = Field(default=0)
    # 月度周期
    monthly_period_start: datetime | None = Field(default=None)
    monthly_period_end: datetime | None = Field(default=None)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class SubscriptionHistory(SQLModel, table=True):
    """
    Subscription history model for audit trail.

    Records all subscription changes:
    - created: New subscription
    - upgraded: Plan upgrade
    - renewed: Subscription renewal
    - expired: Subscription expiration
    - cancelled: Subscription cancellation
    - migrated: Data migration
    """

    __tablename__ = "subscription_history"

    id: str = Field(default_factory=generate_uuid, primary_key=True)
    user_id: str = Field(foreign_key="user.id", index=True)
    action: str  # "created", "upgraded", "renewed", "expired", "cancelled", "migrated"
    plan_name: str
    start_date: datetime
    end_date: datetime | None = None
    event_metadata: dict = Field(default_factory=dict, sa_column=Column(JSON))
    created_at: datetime = Field(default_factory=datetime.utcnow)


class AdminAuditLog(SQLModel, table=True):
    """
    Admin audit log model for administrative actions.

    Tracks all admin operations for security and compliance:
    - Code generation/management
    - Subscription modifications
    - User management actions
    """

    __tablename__ = "admin_audit_log"

    id: str = Field(default_factory=generate_uuid, primary_key=True)
    admin_user_id: str = Field(foreign_key="user.id", index=True)
    action: str
    resource_type: str = Field(index=True)  # "code", "subscription", etc.
    resource_id: str | None = None
    old_value: dict | None = Field(default=None, sa_column=Column(JSON))
    new_value: dict | None = Field(default=None, sa_column=Column(JSON))
    ip_address: str | None = None
    user_agent: str | None = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
